homeassistant/util/logging.py

```diff
@@ -119,28 +119,28 @@ def log_exception(format_err: Callable[..., Any], *args: Any) -> None:
 @overload
 def catch_log_exception(
     func: Callable[..., Coroutine[Any, Any, Any]], format_err: Callable[..., Any]
-) -> Callable[..., Coroutine[Any, Any, None]]:
+) -> Callable[..., Coroutine[Any, Any, None]] | Callable[..., None]:
     """Overload for Callables that return a Coroutine."""
 
 
 @overload
 def catch_log_exception(
     func: Callable[..., Any], format_err: Callable[..., Any]
-) -> Callable[..., None | Coroutine[Any, Any, None]]:
+) -> Callable[..., Coroutine[Any, Any, None]] | Callable[..., None]:
     """Overload for Callables that return Any."""
```

Comment:
Right, not sure how to type the homeassistant @callback type here. `CALLBACK_TYPE` cannot be used here.

Reply:
> Aren't the wanted overloads:
> -> Home Assistant "@callback" in -> @callback out
> -> Coroutine in -> Coroutine out

Not just HA callbacks, any Callable which isn't a coroutine-function, i.e. `Callable[..., Any]`.

The challenge is that which the current Python typing systems it isn't possible to say "Any except Coroutines". Since `Any` includes `Coroutine`, the return type must also include the `Callable[..., Coroutine[Any, Any, Any]]` part.

In summary, this should be the full typing
```py
@overload
def catch_log_exception(
    func: Callable[..., Coroutine[Any, Any, Any]], format_err: Callable[..., Any]
) -> Callable[..., Coroutine[Any, Any, None]]:
    ...

@overload
def catch_log_exception(
    func: Callable[..., Any], format_err: Callable[..., Any]
) -> Callable[..., None] | Callable[..., Coroutine[Any, Any, None]]:
    ...

def catch_log_exception(
    func: Callable[..., Any], format_err: Callable[..., Any]
) -> Callable[..., None] | Callable[..., Coroutine[Any, Any, None]]:
    # ...